atlasdb-conjure/src/main/java/com/palantir/atlasdb/http/v2/CachedComposingSupplier.java

```diff
@@ -0,0 +1,62 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ */
+
+package com.palantir.atlasdb.http.v2;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Objects;
+import java.util.function.Function;
+import java.util.function.Supplier;
+
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import com.palantir.logsafe.UnsafeArg;
+
+/**
+ * Represents a value which is computed by lazily applying the given mapping function to
+ * the result of an 'input' supplier. If the input hasn't changed, we avoid re-running the mapping function (as this
+ * may be expensive).
+ *
+ * <p>This is effectively just Witchcraft's {@code Refreshable} class.
+ */
+final class CachedComposingSupplier<T, U> implements Supplier<U> {
+    private static final Logger log = LoggerFactory.getLogger(CachedComposingSupplier.class);
+    private final Supplier<T> inputSupplier;
+    private final Function<T, U> outputFunction;
+
+    private T cachedInput = null;
+    private U cachedOutput = null;
+
+    CachedComposingSupplier(Supplier<T> inputSupplier, Function<T, U> outputFunction) {
+        this.inputSupplier = inputSupplier;
+        this.outputFunction = outputFunction;
+    }
+
+    public <V> CachedComposingSupplier<U, V> map(Function<U, V> map) {
+        return new CachedComposingSupplier<>(this, map);
+    }
+
+    @Override
+    public synchronized U get() {
```

Comment:
I put together a lock-free version of this because I was concerned about the SupplierChannel hitting a `synchronized` block: https://internal-github/foundry/aaaa-client-java/pull/910/files

Reply:
haha nice thanks! Carter and I were chatting about pulling out Refreshable so that we can get rid of all of these approximations

Reply:
amusingly it seems like we did in fact start this process: https://github.com/palantir/refreshable

Reply:
Great, can we get that in, would love to get rid of all the permutations of this code.

Reply:
Sure, i'll update to use a refreshable